File: minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.cpp


class Solution {
    bool solve(int mid,vector<int>&nums,int m,int k){
        int count=0;
        for(int i=0;i<nums.size();i++){
            if(nums[i]<=mid){
                count++;
            }
            else
            {
                count=0;
            }
            if(count==k){
                m--;
                count=0;
            }
            if(m==0){
                return true;
            }
        }
        return false;
    }
public:
    int minDays(vector<int>& nums, int m, int k) {
        
        int lo=1e6,hi=0,n=nums.size();
        if(m*k>n){
            return -1;
        }
        for(int i=0;i<n;i++){
            lo=min(lo,nums[i]);
            hi=max(hi,nums[i]);
        }
        
        while(lo<hi){
        
            int mid=lo+(hi-lo)/2;

            if(solve(mid,nums,m,k))                  
                hi=mid;
            else
                lo=mid+1;
        }
        return lo;
    }
};

File: 1605-minimum-number-of-days-to-make-m-bouquets/minimum-number-of-days-to-make-m-bouquets.py
class Solution:
    def minDays(self, nums: List[int], m: int, k: int) -> int:
        def solve(mid):
            count = 0
            count_k = 0
            for i in range(len(nums)):
                if nums[i]<=mid:
                    count+=1
                    if count == k:
                        count = 0
                        count_k += 1
                else:
                    count = 0    
            return count_k>=m
        lo = min(nums)
        hi = max(nums)
        ans = -1
        while lo<=hi:
            mid = lo+(hi-lo)//2
            if solve(mid):
                ans = mid
                hi = mid - 1
            else:
                lo = mid + 1
        return ans 
# if(solve(mid)){
#             ans=mid;
#             r=mid-1;
#         }
#         else{
#             l=mid+1;
#         }